libopenage/screenshot.cpp

```diff
@@ -46,44 +50,83 @@ std::string ScreenshotManager::gen_next_filename() {
 
 
 void ScreenshotManager::save_screenshot() {
-	std::string filename = this->gen_next_filename();
+	coord::pixel_t width = this->window_size.x,
+	               height = this->window_size.y;
 
-	log::log(MSG(info) << "Saving screenshot to " << filename);
+	std::shared_ptr<uint8_t> pxdata(new uint8_t[4*width*height], std::default_delete<uint8_t[]>());
+	glReadPixels(0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, pxdata.get());
 
-	int32_t rmask, gmask, bmask, amask;
-	rmask = 0x000000FF;
-	gmask = 0x0000FF00;
-	bmask = 0x00FF0000;
-	amask = 0xFF000000;
+	auto encode_function = [this, pxdata, width, height] () mutable {
+		return this->encode_png(pxdata, width, height);
+	};
+	this->job_manager->enqueue<bool>(encode_function);
+}
 
-	SDL_Surface *screen = SDL_CreateRGBSurface(
-	SDL_SWSURFACE,
-	this->window_size.x,
-	this->window_size.y,
-	32,
-	rmask, gmask, bmask, amask);
 
-	size_t pxcount = screen->w * screen->h;
+bool ScreenshotManager::encode_png(std::shared_ptr<uint8_t> pxdata,
+	                               coord::pixel_t width, coord::pixel_t height) {
+	std::string filename = this->gen_next_filename();
+	FILE *fout = fopen(filename.c_str(), "wb");
+	if (fout == NULL) {
+		log::log(MSG(err) << "Could not open '"<< filename << "': "
+		         << std::string(strerror(errno)));
+		return false;
+	}
 
-	auto pxdata = std::make_unique<uint32_t[]>(pxcount);
+	auto warn_fn = [] (png_structp /*png_ptr*/, png_const_charp message) {
+		log::log(MSG(err) << "Creating screenshot failed: libpng error: " << message);
+	};
+	auto err_fn = [] (png_structp png_ptr, png_const_charp message) {
+		log::log(MSG(err) << "Creating screenshot failed: libpng error: " << message);
+		longjmp(png_jmpbuf(png_ptr), 1);
+	};
+
+
+	png_structp png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING,
+	                                              (png_voidp) NULL,
+	                                              err_fn, warn_fn);
+	if (!png_ptr) {
+		fclose(fout);
+		return false;
+	}
 
-	glReadPixels(0, 0,
-	             this->window_size.x, this->window_size.y,
-	             GL_RGBA, GL_UNSIGNED_BYTE, pxdata.get());
+	png_infop info_ptr = png_create_info_struct(png_ptr);
+	if (!info_ptr) {
+		fclose(fout);
+		png_destroy_write_struct(&png_ptr, (png_infopp) NULL);
+		return false;
+	}
 
-	uint32_t *surface_pxls = (uint32_t *)screen->pixels;
+	if (setjmp(png_jmpbuf(png_ptr))) {
```

Comment:
After I recovered from the shock reading this, I read upon the code of libpng - It is actually the only way to go without terminating the whole game...

Reply:
they were inspired by c++ exceptions.

Reply:
http://thread.gmane.org/gmane.comp.graphics.png.devel/5641
The mentioned callbacks are here: http://www.libpng.org/pub/png/libpng-manual.txt
found under "Setting up callback code".
If we let libpng handle the allocations/frees, this is also not necessary. This really seems strange and we should look into how Pillow or SDL2_Image interacts with libpng.

Reply:
https://hg.libsdl.org/SDL_image/file/6332f9425dcc/IMG_png.c#l388 They also do it, but do different handling for libpng12.

On the CI there is libpng16 (archlinux). libpng12 is very old, so I think we should target >12.

Reply:
so leave it as it is?

Reply:
on your first comment: 

> If we let libpng handle the allocations/frees, this is also not necessary.

Did you mean: `If we handle the allocations/frees, this is also not necessary.` ?
this is what the 4th post on gmane says.